src/Microsoft.AspNetCore.Server.Kestrel.Libuv/Internal/Http/Connection.cs

```diff
@@ -0,0 +1,281 @@
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.IO.Pipelines;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Server.Kestrel.Adapter;
+using Microsoft.AspNetCore.Server.Kestrel.Adapter.Internal;
+using Microsoft.AspNetCore.Server.Kestrel.Internal.Infrastructure;
+using Microsoft.AspNetCore.Server.Kestrel.Internal.Networking;
+using Microsoft.AspNetCore.Server.Kestrel.Transport;
+using Microsoft.AspNetCore.Server.Kestrel.Transport.Exceptions;
+using Microsoft.Extensions.Internal;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Internal.Http
+{
+    public class Connection : ConnectionContext, ITimeoutControl
+    {
+        private const int MinAllocBufferSize = 2048;
+
+        private static readonly Action<UvStreamHandle, int, object> _readCallback =
+            (handle, status, state) => ReadCallback(handle, status, state);
+        
+        private static readonly Func<UvStreamHandle, int, object, LibuvFunctions.uv_buf_t> _allocCallback =
+            (handle, suggestedsize, state) => AllocCallback(handle, suggestedsize, state);
+
+        private readonly UvStreamHandle _socket;
+        private IConnectionContext _connectionContext;
+
+        private TaskCompletionSource<object> _socketClosedTcs = new TaskCompletionSource<object>();
+
+        private long _lastTimestamp;
+        private long _timeoutTimestamp = long.MaxValue;
+        private TimeoutAction _timeoutAction;
+        private WritableBuffer? _currentWritableBuffer;
+
+        public Connection(ListenerContext context, UvStreamHandle socket) : base(context)
+        {
+            _socket = socket;
+            socket.Connection = this;
+            TimeoutControl = this;
+
+            var tcpHandle = _socket as UvTcpHandle;
+            if (tcpHandle != null)
+            {
+                RemoteEndPoint = tcpHandle.GetPeerIPEndPoint();
+                LocalEndPoint = tcpHandle.GetSockIPEndPoint();
+            }
+
+            _lastTimestamp = Thread.Loop.Now();
+        }
+
+        // For testing
+        public Connection()
+        {
+        }
+
+        public string ConnectionId { get; set; }
+        public IPipeWriter Input { get; set; }
+        public SocketOutputConsumer Output { get; set; }
+
+        private IKestrelTrace Log => ListenerContext.TransportContext.Log;
+        private IConnectionHandler ConnectionHandler => ListenerContext.TransportContext.ConnectionHandler;
+        private KestrelThread Thread => ListenerContext.Thread;
+
+        public void Start()
+        {
+            try
+            {
+                _connectionContext = ConnectionHandler.OnConnection(this, Thread, Thread);
+                ConnectionId = _connectionContext.ConnectionId;
+
+                Log.ConnectionStart(ConnectionId);
+                KestrelEventSource.Log.ConnectionStart(this);
```

Comment:
Agreed. I had

- Re-enable KestrelEventSource

in the original description, because i had it commented out for a while until I could fix it. I might file an issue to do this after merged if not trivial.